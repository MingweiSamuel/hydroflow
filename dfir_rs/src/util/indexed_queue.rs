//! A priority queue where the priorities are indices (small non-negative integers).

use std::collections::VecDeque;

/// A priority queue where the priorities are indices (small non-negative integers).
///
/// Larger indices are higher priority.
pub struct IndexedQueue<T> {
    /// Invariant: the last `VecDeque` is non-empty.
    queues: Vec<VecDeque<T>>,
}
impl<T> Default for IndexedQueue<T> {
    fn default() -> Self {
        Self { queues: Vec::new() }
    }
}
impl<T> IndexedQueue<T> {
    /// Creates a new `IndexedQueue`.
    pub fn new() -> Self {
        Self::default()
    }

    /// Pushes an element to the queue with the given index.
    pub fn push(&mut self, index: usize, element: T) {
        if index >= self.queues.len() {
            self.queues.resize_with(index + 1, Default::default);
        }
        self.queues[index].push_back(element);
    }

    /// Pops an element from the queue with the largest index.
    pub fn pop(&mut self) -> Option<(usize, T)> {
        let item = self.queues.last_mut()?.pop_front().unwrap();
        let index = self.queues.len() - 1;
        // Remove trailing empty queues to maintain the invariant.
        while self.queues.last().is_some_and(VecDeque::is_empty) {
            self.queues.pop();
        }
        Some((index, item))
    }

    /// Returns the number of elements in the `IndexedQueue`.
    pub fn len(&self) -> usize {
        self.queues.iter().map(VecDeque::len).sum()
    }

    /// Returns `true` if the `IndexedQueue` is empty.
    pub fn is_empty(&self) -> bool {
        self.queues.is_empty()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_indexed_queue() {
        let mut queue = IndexedQueue::new();
        assert_eq!(queue.len(), 0);
        assert!(queue.is_empty());

        queue.push(0, 0);
        queue.push(1, 1);
        queue.push(0, 2);
        queue.push(5, 3);
        queue.push(1, 4);

        assert_eq!(queue.len(), 5);
        assert!(!queue.is_empty());

        assert_eq!(queue.pop(), Some((5, 3)));

        assert_eq!(queue.len(), 4);
        assert!(!queue.is_empty());

        assert_eq!(queue.pop(), Some((1, 1)));
        assert_eq!(queue.pop(), Some((1, 4)));
        assert_eq!(queue.pop(), Some((0, 0)));

        assert_eq!(queue.len(), 1);
        assert!(!queue.is_empty());

        assert_eq!(queue.pop(), Some((0, 2)));
        assert_eq!(queue.pop(), None);

        assert_eq!(queue.len(), 0);
        assert!(queue.is_empty());
    }
}
