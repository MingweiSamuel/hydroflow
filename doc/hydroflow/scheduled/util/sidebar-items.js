initSidebarItems({"enum":[["Once","The receiving half of a once channel."]],"fn":[["once","Creates a once channel where the [SendOnce] end can set a value which the [Once] end will receive and store."]],"struct":[["SendOnce","The sending half of a once channel."]]});