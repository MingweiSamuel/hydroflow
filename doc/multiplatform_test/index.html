<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`multiplatform_test`"><title>multiplatform_test - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-a23c9d150ad85d8b.css" id="mainThemeStyle"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="multiplatform_test" data-themes="" data-resource-suffix="" data-rustdoc-version="1.72.0-nightly (330727467 2023-06-29)" data-channel="nightly" data-search-js="search-91b58767224854cb.js" data-settings-js="settings-de11bff964e9d4e5.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ><script src="../static.files/storage-59fd9b8ccb335783.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-0795b7d26be81095.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../multiplatform_test/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../multiplatform_test/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate multiplatform_test</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 0.0.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">multiplatform_test</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/multiplatform_test/lib.rs.html#1-240">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="multiplatform_test"><a href="#multiplatform_test"><code>multiplatform_test</code></a></h2>
<p>Provides a proc-macro that expands to testing on platforms relevant to
Hydroflow. By default, expands to testing on the host (using the normal
<code>#[test]</code> attribute) and wasm (using <code>#[wasm_bindgen_test]</code>).</p>
<p>For example, the test</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[multiplatform_test]
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
<p>Expands to</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[test]
#[wasm_bindgen_test::wasm_bindgen_test]
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
<h3 id="installation"><a href="#installation">Installation</a></h3><div class="example-wrap"><pre class="language-toml"><code>[dependencies]
multiplatform_test = * # replace with version.
</code></pre></div>
<p>If you’re using <code>wasm</code> naturally you will need to add the <a href="https://crates.io/crates/wasm-bindgen-test-macro/"><code>wasm_bindgen_test</code></a>
crate as a dependency.</p>
<h3 id="usage"><a href="#usage">Usage</a></h3><h4 id="specifying-platforms"><a href="#specifying-platforms">Specifying platforms</a></h4>
<p>There are many platforms which can be specified:</p>
<ul>
<li><code>test</code> - Adds a standard <a href="https://doc.rust-lang.org/reference/attributes/testing.html#the-test-attribute"><code>#[test]</code> attribute</a>.</li>
<li><code>tokio</code> - Adds a <a href="https://docs.rs/tokio/latest/tokio/attr.test.html"><code>#[tokio::test]</code> attribute</a>.</li>
<li><code>async_std</code> - Adds an <a href="https://docs.rs/async-std/latest/async_std/attr.test.html"><code>#[async_std::test]</code> attribute</a>.</li>
<li><code>hydroflow</code> - Adds a <a href="https://docs.rs/hydroflow/latest/hydroflow/attr.test.html"><code>#[hydroflow::test]</code> attribute</a>.</li>
<li><code>wasm</code> - Adds a <a href="https://docs.rs/wasm-bindgen-test/0.3.36/wasm_bindgen_test/attr.wasm_bindgen_test.html"><code>#[wasm_bindgen_test::wasm_bindgen_test]</code> attribute</a>.</li>
<li><code>env_logging</code> - Registers <a href="https://docs.rs/env_logger/latest/env_logger/"><code>env_logger</code></a> for <a href="https://docs.rs/log/latest/log/"><code>log</code>ging</a>.</li>
<li><code>env_tracing</code> - Registers a <a href="https://docs.rs/tracing-subscriber/latest/tracing_subscriber/index.html#reexport.FmtSubscriber"><code>FmtSubscriber</code></a> with an <a href="https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html"><code>EnvFilter</code></a> for <a href="https://docs.rs/tracing/latest/tracing/"><code>tracing</code></a>.</li>
</ul>
<p>You can test on a subset of platforms by passing in the platforms in parens:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[multiplatform_test(test)]  </span><span class="comment">// Only test on the standard `#[test]` platform, but enables logging
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
<p>expands to</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>multiplatform_test::multiplatform_test;

<span class="attr">#[test]
</span><span class="kw">fn </span>my_test() {
  <span class="comment">// ...
</span>}</code></pre></div>
</div></details><h2 id="attributes" class="small-section-header"><a href="#attributes">Attribute Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.multiplatform_test.html" title="attr multiplatform_test::multiplatform_test">multiplatform_test</a></div><div class="desc docblock-short">See the <a href="index.html" title="mod multiplatform_test">crate</a> docs for usage information.</div></li></ul></section></div></main></body></html>