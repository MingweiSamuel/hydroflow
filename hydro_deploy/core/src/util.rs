use std::io;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use anyhow::Result;
use async_channel::Sender;
use futures::future::join_all;
use futures::{Future, StreamExt};
use futures_core::Stream;

pub async fn async_retry<T, F: Future<Output = Result<T>>>(
    mut thunk: impl FnMut() -> F,
    count: usize,
    delay: Duration,
) -> Result<T> {
    for _ in 1..count {
        let result = thunk().await;
        if result.is_ok() {
            return result;
        } else {
            tokio::time::sleep(delay).await;
        }
    }

    thunk().await
}

type PriorityBroadcacst = (
    Arc<Mutex<Option<tokio::sync::oneshot::Sender<String>>>>,
    Arc<Mutex<Vec<Sender<String>>>>,
);

pub fn prioritized_broadcast<T: Stream<Item = io::Result<String>> + Send + Unpin + 'static>(
    mut lines: T,
    default: impl Fn(String) + Send + 'static,
) -> PriorityBroadcacst {
    let priority_receivers = Arc::new(Mutex::new(None::<tokio::sync::oneshot::Sender<String>>));
    let receivers = Arc::new(Mutex::new(Vec::<Sender<String>>::new()));

    let weak_priority_receivers = Arc::downgrade(&priority_receivers);
    let weak_receivers = Arc::downgrade(&receivers);

    tokio::spawn(async move {
        while let Some(Result::Ok(line)) = lines.next().await {
            if let Some(cli_receivers) = weak_priority_receivers.upgrade() {
                let mut cli_receivers = cli_receivers.lock().unwrap();

                let successful_send = if let Some(r) = cli_receivers.take() {
                    r.send(line.clone()).is_ok()
                } else {
                    false
                };
                drop(cli_receivers);

                if successful_send {
                    continue;
                }
            }

            if let Some(receivers) = weak_receivers.upgrade() {
                let send_all = {
                    let mut receivers = receivers.lock().unwrap();
                    receivers.retain(|receiver| !receiver.is_closed());
                    join_all(receivers.iter().map(|receiver| {
                        // Create a future which doesn't need to hold the `receivers` lock.
                        let receiver = receiver.clone();
                        let line = &line;
                        async move { receiver.send(line.clone()).await }
                    }))
                    // Do not `.await` while holding onto the `std::sync::Mutex` `receivers` lock.
                };

                let successful_send = send_all.await.into_iter().any(|result| result.is_ok());
                if !successful_send {
                    (default)(line);
                }
            } else {
                break;
            }
        }

        if let Some(cli_receivers) = weak_priority_receivers.upgrade() {
            let mut cli_receivers = cli_receivers.lock().unwrap();
            drop(cli_receivers.take());
        }

        if let Some(receivers) = weak_receivers.upgrade() {
            let mut receivers = receivers.lock().unwrap();
            receivers.clear();
        }
    });

    (priority_receivers, receivers)
}

#[cfg(test)]
mod test {
    use futures::StreamExt;

    #[tokio::test]
    async fn broadcast_listeners_close_when_source_does() {
        let (tx, rx) = async_channel::unbounded::<_>();
        let (_, receivers) = super::prioritized_broadcast(rx, |_| {});

        let (tx2, mut rx2) = async_channel::unbounded::<_>();

        receivers.lock().unwrap().push(tx2);

        tx.send(Ok("hello".to_string())).await.unwrap();
        assert_eq!(rx2.next().await, Some("hello".to_string()));

        let wait_again = tokio::spawn(async move { rx2.next().await });

        drop(tx);

        assert_eq!(wait_again.await.unwrap(), None);
    }
}
