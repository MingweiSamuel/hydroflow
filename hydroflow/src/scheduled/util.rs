use std::rc::Rc;

use once_cell::unsync::OnceCell;

/**
 * Creates a once channel where the [SendOnce] end can set a value which the [Once] end will receive and store.
 */
pub fn once<T>() -> (SendOnce<T>, Once<T>) {
    let channel = Rc::new(OnceCell::new());
    let sender = SendOnce {
        channel: channel.clone(),
    };
    let once = Once { channel };
    (sender, once)
}

/**
 * The sending half of a once channel.
 */
pub struct SendOnce<T> {
    channel: Rc<OnceCell<T>>,
}
impl<T> SendOnce<T> {
    pub fn send(self, item: T) {
        let result = self.channel.set(item);
        assert!(result.is_ok());
    }
}

/**
 * The receiving half of a once channel.
 */
pub struct Once<T> {
    channel: Rc<OnceCell<T>>,
}
impl<T> Once<T> {
    pub fn get(&mut self) -> &mut T {
        Rc::get_mut(&mut self.channel)
            .expect("Called Once::get() before value is set.")
            .get_mut()
            .unwrap()
    }
}
