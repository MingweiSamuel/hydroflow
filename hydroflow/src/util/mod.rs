pub mod slot_map;
