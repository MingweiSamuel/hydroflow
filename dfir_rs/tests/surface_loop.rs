use dfir_rs::util::iter_batches_stream;
use dfir_rs::{assert_graphvis_snapshots, dfir_syntax};
use multiplatform_test::multiplatform_test;

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_flo_syntax() {
    let mut df = dfir_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..12, 3));
        loop {
            // TODO(mingwei): cross_join type negotion should allow us to eliminate `flatten()`.
            users -> batch() -> flatten() -> [0]cp;
            messages -> batch() -> flatten() -> [1]cp;
            cp = cross_join::<'static, 'tick>()
                -> map(|item| (context.current_tick().0, item))
                -> assert_eq([
                    (0, ("alice", 0)),
                    (0, ("alice", 1)),
                    (0, ("alice", 2)),
                    (0, ("bob", 0)),
                    (0, ("bob", 1)),
                    (0, ("bob", 2)),
                    (1, ("alice", 3)),
                    (1, ("alice", 4)),
                    (1, ("alice", 5)),
                    (1, ("bob", 3)),
                    (1, ("bob", 4)),
                    (1, ("bob", 5)),
                    (2, ("alice", 6)),
                    (2, ("alice", 7)),
                    (2, ("alice", 8)),
                    (2, ("bob", 6)),
                    (2, ("bob", 7)),
                    (2, ("bob", 8)),
                    (3, ("alice", 9)),
                    (3, ("alice", 10)),
                    (3, ("alice", 11)),
                    (3, ("bob", 9)),
                    (3, ("bob", 10)),
                    (3, ("bob", 11)),
                ]);
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_flo_nested() {
    let mut df = dfir_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..12, 3));
        loop {
            // TODO(mingwei): cross_join type negotion should allow us to eliminate `flatten()`.
            users -> batch() -> flatten() -> [0]cp;
            messages -> batch() -> flatten() -> [1]cp;
            cp = cross_join::<'static, 'tick>();
            loop {
                cp
                    -> all_once()
                    -> map(|vec| (context.current_tick().0, vec))
                    -> assert_eq([
                        (0, vec![("alice", 0), ("alice", 1), ("alice", 2), ("bob", 0), ("bob", 1), ("bob", 2)]),
                        (1, vec![("alice", 3), ("alice", 4), ("alice", 5), ("bob", 3), ("bob", 4), ("bob", 5)]),
                        (2, vec![("alice", 6), ("alice", 7), ("alice", 8), ("bob", 6), ("bob", 7), ("bob", 8)]),
                        (3, vec![("alice", 9), ("alice", 10), ("alice", 11), ("bob", 9), ("bob", 10), ("bob", 11)]),
                    ]);
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}

#[multiplatform_test]
pub fn test_flo_repeat_n() {
    let mut df = dfir_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..12, 3));
        loop {
            // TODO(mingwei): cross_join type negotion should allow us to eliminate `flatten()`.
            users -> batch() -> flatten() -> [0]cp;
            messages -> batch() -> flatten() -> [1]cp;
            cp = cross_join::<'static, 'tick>();
            loop {
                cp
                    -> repeat_n(3)
                    -> map(|vec| (context.current_tick().0, vec))
                    -> inspect(|x| println!("{:?}", x))
                    -> assert_eq([
                        (0, vec![("alice", 0), ("alice", 1), ("alice", 2), ("bob", 0), ("bob", 1), ("bob", 2)]),
                        (0, vec![("alice", 0), ("alice", 1), ("alice", 2), ("bob", 0), ("bob", 1), ("bob", 2)]),
                        (0, vec![("alice", 0), ("alice", 1), ("alice", 2), ("bob", 0), ("bob", 1), ("bob", 2)]),
                        (1, vec![("alice", 3), ("alice", 4), ("alice", 5), ("bob", 3), ("bob", 4), ("bob", 5)]),
                        (1, vec![("alice", 3), ("alice", 4), ("alice", 5), ("bob", 3), ("bob", 4), ("bob", 5)]),
                        (1, vec![("alice", 3), ("alice", 4), ("alice", 5), ("bob", 3), ("bob", 4), ("bob", 5)]),
                        (2, vec![("alice", 6), ("alice", 7), ("alice", 8), ("bob", 6), ("bob", 7), ("bob", 8)]),
                        (2, vec![("alice", 6), ("alice", 7), ("alice", 8), ("bob", 6), ("bob", 7), ("bob", 8)]),
                        (2, vec![("alice", 6), ("alice", 7), ("alice", 8), ("bob", 6), ("bob", 7), ("bob", 8)]),
                        (3, vec![("alice", 9), ("alice", 10), ("alice", 11), ("bob", 9), ("bob", 10), ("bob", 11)]),
                        (3, vec![("alice", 9), ("alice", 10), ("alice", 11), ("bob", 9), ("bob", 10), ("bob", 11)]),
                        (3, vec![("alice", 9), ("alice", 10), ("alice", 11), ("bob", 9), ("bob", 10), ("bob", 11)]),
                    ]);
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_flo_repeat_n_nested() {
    let mut df = dfir_syntax! {
        usrs1 = source_iter(["alice", "bob"]);
        loop {
            usrs2 = usrs1 -> batch() -> flatten();
            loop {
                usrs3 = usrs2 -> repeat_n(3) -> flatten()
                    -> inspect(|x| println!("{:?} {}", x, context.is_first_loop_iteration()));
                loop {
                    usrs3 -> repeat_n(3)
                        -> inspect(|x| println!("{:?} {}", x, context.is_first_loop_iteration()))
                        -> assert_eq([
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                        ]);
                }
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}

#[multiplatform_test]
pub fn test_flo_repeat_n_multiple_nested() {
    let mut df = dfir_syntax! {
        usrs1 = source_iter(["alice", "bob"]);
        loop {
            usrs2 = usrs1 -> batch() -> flatten();
            loop {
                usrs3 = usrs2 -> repeat_n(3) -> flatten()
                    -> inspect(|x| println!("{:?} {}", x, context.is_first_loop_iteration()))
                    -> tee();
                loop {
                    usrs3 -> repeat_n(3)
                        -> inspect(|x| println!("{} {:?} {}", line!(), x, context.is_first_loop_iteration()))
                        -> assert_eq([
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                        ]);
                }
                loop {
                    usrs3 -> repeat_n(3)
                        -> inspect(|x| println!("{} {:?} {}", line!(), x, context.is_first_loop_iteration()))
                        -> assert_eq([
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                            vec!["alice", "bob"],
                        ]);
                }
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}

#[multiplatform_test]
pub fn test_flo_repeat_kmeans() {
    const POINTS: &[[i32; 2]] = &[
        [-210, -104],
        [-226, -143],
        [-258, -119],
        [-331, -129],
        [-250, -69],
        [-202, -113],
        [-222, -133],
        [-232, -155],
        [-220, -107],
        [-159, -109],
        [-49, 57],
        [-156, 52],
        [-22, 125],
        [-140, 168],
        [-118, 89],
        [-93, 133],
        [-101, 80],
        [-145, 79],
        [187, 36],
        [208, -66],
        [142, 5],
        [232, 41],
        [91, -37],
        [132, 16],
        [248, -39],
        [158, 65],
        [108, -41],
        [171, -121],
        [147, 5],
        [192, 58],
    ];
    const CENTROIDS: &[[i32; 2]] = &[[-50, 0], [0, 0], [50, 0]];

    let mut df = dfir_syntax! {
        init_points = source_iter(POINTS) -> map(std::clone::Clone::clone);
        init_centroids = source_iter(CENTROIDS) -> map(std::clone::Clone::clone);
        loop {
            batch_points = init_points -> batch() -> inspect(|x| println!("BATCH POINTS {} {:?}", line!(), x)) -> flatten();
            batch_centroids = init_centroids -> batch() -> inspect(|x| println!("BATCH CENTROIDS {} {:?}", line!(), x)) -> flatten();

            loop {
                points = batch_points
                    -> repeat_n(10)
                    -> inspect(|x| println!("CJ0 {} {:?}", line!(), x))
                    -> flatten()
                    -> [0]cj;
                batch_centroids -> all_once() -> flatten() -> centroids;

                centroids = union()
                    -> inspect(|x| println!("CJ1 {} {:?}", line!(), x))
                    -> [1]cj;

                cj = cross_join_multiset()
                    -> inspect(|x| println!("cjm: {:?}", x))
                    -> map(|(point, centroid): ([i32; 2], [i32; 2])| {
                        let dist2 = (point[0] - centroid[0]).pow(2) + (point[1] - centroid[1]).pow(2);
                        (point, (dist2, centroid))
                    })
                    -> reduce_keyed(|(a_dist2, a_centroid), (b_dist2, b_centroid)| {
                        if b_dist2 < *a_dist2 {
                            *a_dist2 = b_dist2;
                            *a_centroid = b_centroid;
                        }
                    })
                    -> map(|(point, (_dist2, centroid))| {
                        (centroid, (point, 1))
                    })
                    -> reduce_keyed(|(p1, n1), (p2, n2): ([i32; 2], i32)| {
                        p1[0] += p2[0];
                        p1[1] += p2[1];
                        *n1 += n2;
                    })
                    -> map(|(_centroid, (p, n))| {
                         [p[0] / n, p[1] / n]
                    })
                    -> next_loop()
                    -> inspect(|x| println!("centroid: {:?}", x))
                    -> centroids;
            }
        }
    };
    assert_graphvis_snapshots!(df);
    df.run_available();
}
