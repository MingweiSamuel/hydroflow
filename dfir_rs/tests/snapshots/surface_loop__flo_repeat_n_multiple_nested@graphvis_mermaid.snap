---
source: dfir_rs/tests/surface_loop.rs
expression: "df.meta_graph().unwrap().to_mermaid(& Default :: default())"
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
classDef otherClass fill:#fdc,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa
1v1[\"(1v1) <code>source_iter([&quot;alice&quot;, &quot;bob&quot;])</code>"/]:::pullClass
2v1[\"(2v1) <code>batch()</code>"/]:::pullClass
3v1[\"(3v1) <code>flatten()</code>"/]:::pullClass
4v1[\"(4v1) <code>repeat_n(3)</code>"/]:::pullClass
5v1[\"(5v1) <code>flatten()</code>"/]:::pullClass
6v1[\"(6v1) <code>inspect(|x| println!(&quot;{:?} {}&quot;, x, context.is_first_loop_iteration()))</code>"/]:::pullClass
7v1[/"(7v1) <code>tee()</code>"\]:::pushClass
8v1[\"(8v1) <code>repeat_n(3)</code>"/]:::pullClass
9v1[\"(9v1) <code>inspect(|x| println!(&quot;{} {:?} {}&quot;, line!(), x, context.is_first_loop_iteration()))</code>"/]:::pullClass
10v1[/"<div style=text-align:center>(10v1)</div> <code>assert_eq([<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>])</code>"\]:::pushClass
11v1[\"(11v1) <code>repeat_n(3)</code>"/]:::pullClass
12v1[\"(12v1) <code>inspect(|x| println!(&quot;{} {:?} {}&quot;, line!(), x, context.is_first_loop_iteration()))</code>"/]:::pullClass
13v1[/"<div style=text-align:center>(13v1)</div> <code>assert_eq([<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>    vec![&quot;alice&quot;, &quot;bob&quot;],<br>])</code>"\]:::pushClass
14v1["(14v1) <code>handoff</code>"]:::otherClass
15v1["(15v1) <code>handoff</code>"]:::otherClass
16v1["(16v1) <code>handoff</code>"]:::otherClass
17v1["(17v1) <code>handoff</code>"]:::otherClass
2v1-->3v1
1v1-->14v1
6v1-->7v1
5v1-->6v1
4v1-->5v1
3v1-->15v1
9v1-->10v1
8v1-->9v1
7v1-->16v1
12v1-->13v1
11v1-->12v1
7v1-->17v1
14v1-->2v1
15v1--x4v1; linkStyle 13 stroke:red
16v1--x8v1; linkStyle 14 stroke:red
17v1--x11v1; linkStyle 15 stroke:red
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    1v1
    subgraph sg_1v1_var_usrs1 ["var <tt>usrs1</tt>"]
        1v1
    end
end
subgraph sg_2v1 ["sg_2v1 stratum 0"]
    2v1
    3v1
    subgraph sg_2v1_var_usrs2 ["var <tt>usrs2</tt>"]
        2v1
        3v1
    end
end
subgraph sg_3v1 ["sg_3v1 stratum 0"]
    4v1
    5v1
    6v1
    7v1
    subgraph sg_3v1_var_usrs3 ["var <tt>usrs3</tt>"]
        4v1
        5v1
        6v1
        7v1
    end
end
subgraph sg_4v1 ["sg_4v1 stratum 0"]
    8v1
    9v1
    10v1
end
subgraph sg_5v1 ["sg_5v1 stratum 0"]
    11v1
    12v1
    13v1
end
